crates/typst-syntax/src/parser.rs

```diff
@@ -526,52 +526,44 @@ fn math_arg<'s>(p: &mut Parser<'s>, seen: &mut FxHashSet<&'s str>) -> bool {
     let m = p.marker();
     let start = p.current_start();
 
-    if p.at(SyntaxKind::Dot) {
-        // Parses a spread argument: `..args`.
-        if let Some(spread) = p.lexer.maybe_math_spread_arg(start) {
-            p.token.node = spread;
-            p.eat();
-            let m_arg = p.marker();
-            // TODO: Refactor to combine with the other call to `math_exprs`.
-            let count =
-                math_exprs(p, syntax_set!(End, Dollar, Comma, Semicolon, RightParen));
-            if count == 0 {
-                let dots = vec![
-                    SyntaxNode::leaf(SyntaxKind::MathText, "."),
-                    SyntaxNode::leaf(SyntaxKind::MathText, "."),
-                ];
-                p[m] = SyntaxNode::inner(SyntaxKind::Math, dots);
-            } else {
-                if count > 1 {
-                    p.wrap(m_arg, SyntaxKind::Math);
-                }
-                p.wrap(m, SyntaxKind::Spread);
-            }
-            return true;
-        }
-    }
+    let mut arg_kind = None;
 
-    let mut positional = true;
-    if p.at_set(syntax_set!(MathText, MathIdent, Underscore)) {
+    if p.at(SyntaxKind::Dot)
+        && let Some(spread) = p.lexer.maybe_math_spread_arg(start)
+    {
+        // Parses a spread argument: `..args`.
+        arg_kind = Some(SyntaxKind::Spread);
+        p.token.node = spread;
+        p.eat();
+    } else if p.at_set(syntax_set!(MathText, MathIdent, Underscore))
+        && let Some(named) = p.lexer.maybe_math_named_arg(start)
+    {
         // Parses a named argument: `thickness: #12pt`.
-        if let Some(named) = p.lexer.maybe_math_named_arg(start) {
-            p.token.node = named;
-            let text = p.current_text();
-            p.eat();
-            p.convert_and_eat(SyntaxKind::Colon);
-            if !seen.insert(text) {
-                p[m].convert_to_error(eco_format!("duplicate argument: {text}"));
-            }
-            positional = false;
+        arg_kind = Some(SyntaxKind::Named);
+        p.token.node = named;
+        let text = p.current_text();
+        p.eat();
+        p.convert_and_eat(SyntaxKind::Colon);
+        if !seen.insert(text) {
+            p[m].convert_to_error(eco_format!("duplicate argument: {text}"));
         }
     }
 
-    // Parses a normal positional argument.
+    // Parses the argument itself.
     let arg = p.marker();
     let count = math_exprs(p, syntax_set!(End, Dollar, Comma, Semicolon, RightParen));
-    if count == 0 {
-        // Named argument requires a value.
-        if !positional {
+
+    if count == 0 && arg_kind == Some(SyntaxKind::Spread) {
+        // A spread arg with no following expression converts back to two dots.
+        arg_kind = None;
+        let dots = vec![
+            SyntaxNode::leaf(SyntaxKind::MathText, "."),
+            SyntaxNode::leaf(SyntaxKind::MathText, "."),
+        ];
+        p[m] = SyntaxNode::inner(SyntaxKind::Math, dots);
```

Comment:
maybe not worth it, but just a thought: if we just push the text nodes flatly, we could maybe unify the `p.wrap(arg, SyntaxKind::Math);` for all branches?